src/mod/ident.mod/ident.c

```diff
@@ -0,0 +1,202 @@
+#include "server.mod/server.h"
+
+#define IDENT_METHOD_OIDENT  0
+#define IDENT_METHOD_BUILTIN 1
+
+static Function *global = NULL, *server_funcs = NULL;;
+
+static int ident_method = IDENT_METHOD_OIDENT;
+
+static tcl_ints identints[] = {
+  {"ident-method", &ident_method, 0},
+  {NULL,           NULL,          0}
+};
+
+static void ident_builtin_off();
+
+static cmd_t ident_raw[] = {
+  {"001", "",   (IntFunc) ident_builtin_off, "ident:001"},
+  {NULL,  NULL, NULL,                        NULL}
+};
+
+static void ident_builtin_off()
+{
+  rem_builtins(H_raw, ident_raw);
+}
+
+static void ident_activity(int idx, char *buf, int len)
+{
+  int s;
+  char buf2[128], *pos;
+  ssize_t i;
+
+  s = answer(dcc[idx].sock, &dcc[idx].sockname, 0, 0);
+  if ((i = read(s, buf2, 64)) < 0) {
+    putlog(LOG_MISC, "*", "Ident error: %s", strerror(errno));
+    return;
+  }
+  buf2[i - 1] = 0;
+  if (!(pos = strpbrk(buf2, "\n\r\\0"))) {
+    putlog(LOG_MISC, "*", "Ident error: couldnt read request");
+    return;
+  } 
+  snprintf(pos, (sizeof buf2) - (pos - buf2), " : USERID : UNIX : %s\n\r", botname);
+  if ((i = write(s, buf2, strlen(buf2) + 1)) < 0) {
+    putlog(LOG_MISC, "*", "Ident error: %s", strerror(errno));
+    return;
+  }
+}
+
+static void ident_display(int idx, char *buf)
+{
+  strcpy(buf, "ident (ready)");
+}
+
+static struct dcc_table DCC_IDENTD = {
+  "IDENTD",
+  DCT_LISTEN,
+  NULL,
+  ident_activity,
+  NULL,
+  NULL,
+  ident_display,
+  NULL,
+  NULL,
+  NULL
+};
+
+static void ident_oidentd()
+{
+  char *home = getenv("HOME");
+  char path[121], buf[(sizeof "global{reply \"\"}") + USERLEN];
+  int nbytes;
+  int fd;
+
+  if (!home) {
+    putlog(LOG_MISC, "*",
+           "Ident error: variable HOME is not in the current environment.");
+    return;
+  }
+  if (snprintf(path, sizeof path, "%s/.oidentd.conf", home) >= sizeof path) {
+    putlog(LOG_MISC, "*", "Ident error: path too long.");
+    return;
+  }
+  if ((fd = open(path, O_WRONLY | O_CREAT | O_TRUNC, S_IRUSR | S_IWUSR | S_IROTH)) < 0) {
+    putlog(LOG_MISC, "*", "Ident error: %s", strerror(errno));
+    return;
+  }
+  nbytes = snprintf(buf, sizeof buf, "global{reply \"%s\"}", botuser);
+  if (write(fd, buf, nbytes) < 0)
+    putlog(LOG_MISC, "*", "Ident error: %s", strerror(errno));
+  close(fd);
+}
+
+static void ident_builtin_on()
+{
+  int idx, port = 113, s;
+
+  idx = new_dcc(&DCC_IDENTD, 0);
+  if (idx < 0) {
+    putlog(LOG_MISC, "*", "Ident error: could not get new dcc.");
+    return;
+  }
+  s = open_listen(&port);
+  if (s == -2) {
+    lostdcc(idx);
+    putlog(LOG_MISC, "*", "Ident error: could not bind socket.");
+    return;
+  } else if (s == -1) {
+    lostdcc(idx);
+    putlog(LOG_MISC, "*", "Ident error: could not get socket.");
+    return;
+  }
+  dcc[idx].sock = s;
+  /* dcc[idx].timeval = now; */
+  strcpy(dcc[idx].nick, "(ident)");
+  
+  add_builtins(H_raw, ident_raw);
+}
+
+static void ident_ident()
+{
+  if (ident_method == IDENT_METHOD_OIDENT)
```

Comment:
A `switch` would work well here.

Reply:
With only 2 checks, is shouldn't make much difference. A question of style. So you like switch more.